### PR TITLE
Add FAT for Faces 4.0 FacesContext.getLifecycle() API

### LINKED ISSUE #20841: Faces 4.0: Add FacesContext#getLifecycle()
Spec issue: `https://github.com/jakartaee/faces/issues/1557`

See: https://balusc.omnifaces.org/2021/11/whats-new-in-faces-40.html

### PR DESCRIPTION
**Pull request by pnicolucci**
fixes #20841

### CODE REVIEW

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
Java 17 build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_kDCSIEUXEe2gRt1V5m8KOw

**Comment by pnicolucci** [comment]
Java 17 build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_kDCSIEUXEe2gRt1V5m8KOw

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
Java 17 build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_FSwlwEVtEe2gRt1V5m8KOw&tab=com.ibm.team.build.web.ui.internal.editors.result.overview.OverviewPage

**Comment by pnicolucci** [comment]
Java 17 build https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_ZZOzEEXhEe2gRt1V5m8KOw

**Comment by pnicolucci** [comment]
Java 17 was successful as well.

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/FacesContextGetLifecycle.war/src/io/openliberty/org/apache/myfaces40/fat/facesContext/getLifecycle/bean/AddPhaseListenerBean.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+public class AddPhaseListenerBean {
+    boolean phaseListenerAdded = false;
+
+    public void addPhaseListener() {
+        FacesContext facesContext = FacesContext.getCurrentInstance();
+
+        if (!phaseListenerAdded) {
+            ProgrammaticallyAddedPhaseListener phaseListener = new ProgrammaticallyAddedPhaseListener();
+            facesContext.getLifecycle().addPhaseListener(phaseListener);
+        }
```
Comment (volosied) [inline dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/FacesContextGetLifecycle.war/src/io/openliberty/org/apache/myfaces40/fat/facesContext/getLifecycle/bean/AddPhaseListenerBean.java]: I know this test only runs once, but should phaseListenerAdded be set to true (if testing manually) in the if block?

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_facesContextGetLifecycleServer/server.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        IBM Corporation - initial API and implementation
+ -->
+<server description="Server for testing the Faces 4.0 FacesContext.getLifecycle() API">
+
+    <include location="../fatTestPorts.xml"/>
+
+    <featureManager>
+        <feature>faces-4.0</feature>
+        <feature>cdi-4.0</feature>
+    </featureManager>
```
Comment (volosied) [inline dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_facesContextGetLifecycleServer/server.xml:18]: Should we create a new server for each test? Or could have a generic faces-cdi server?  Maybe something to follow up on later.

**Review by volosied — APPROVED** [review]
Looks good.

Just to verify my understanding. The listener is added during the invoke application phase, so the listener's first full phase is render phase. This is what you check in the test:  `beforePhase: RENDER_RESPONSE` and `afterPhase: RENDER_RESPONSE`

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/FacesContextGetLifecycle.war/src/io/openliberty/org/apache/myfaces40/fat/facesContext/getLifecycle/bean/AddPhaseListenerBean.java (hunk 1 above)
Comment (pnicolucci) [reply dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/FacesContextGetLifecycle.war/src/io/openliberty/org/apache/myfaces40/fat/facesContext/getLifecycle/bean/AddPhaseListenerBean.java]: Yes, that was the whole point of doing it this way, I forgot it. Nice catch, I'll fix it.

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_facesContextGetLifecycleServer/server.xml (hunk 2 above)
Comment (pnicolucci) [reply dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_facesContextGetLifecycleServer/server.xml:18]: I like having one server per test class, makes finding the logs etc easier. If we have some tests we want to group together we could.

Comment (pnicolucci) [reply dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_facesContextGetLifecycleServer/server.xml:18]: https://github.com/OpenLiberty/open-liberty/wiki/FAT-Tests also recommends this approach: `Use one test class per server `

**Comment by pnicolucci** [comment]
> Looks good.
> 
> Just to verify my understanding. The listener is added during the invoke application phase, so the listener's first full phase is render phase. This is what you check in the test: `beforePhase: RENDER_RESPONSE` and `afterPhase: RENDER_RESPONSE`

Hey, yes, if I were to add a `PhaseListener` in the `faces-config.xml` and then run the test we would get the following output:

> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: RESTORE_VIEW(1)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: RESTORE_VIEW(1)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: RENDER_RESPONSE(6)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.No state saving method defined, assuming default server state saving
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: RENDER_RESPONSE(6)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: RESTORE_VIEW(1)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: RESTORE_VIEW(1)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: APPLY_REQUEST_VALUES(2)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: APPLY_REQUEST_VALUES(2)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: PROCESS_VALIDATIONS(3)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: PROCESS_VALIDATIONS(3)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: UPDATE_MODEL_VALUES(4)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: UPDATE_MODEL_VALUES(4)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: INVOKE_APPLICATION(5)
> PAN: adding phase listener programmatically
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: INVOKE_APPLICATION(5)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: beforePhase: RENDER_RESPONSE(6)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.ProgrammaticPhaseListener: beforePhase: RENDER_RESPONSE(6)
>  SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.ProgrammaticPhaseListener: afterPhase: RENDER_RESPONSE(6)
> SRVE0292I: Servlet Message - [FacesContextGetLifecycle]:.AppPhaseListener: afterPhase: RENDER_RESPONSE(6)

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Pull request merged**